diff --git a/src/CSS.php b/src/CSS.php
index b7e2726..6d31a3d 100644
--- a/src/CSS.php
+++ b/src/CSS.php
@@ -216,7 +216,9 @@ class CSS extends Minify
             // check if current file was not imported previously in the same
             // import chain.
             if (in_array($importPath, $parents)) {
-                throw new FileImportException('Failed to import file "' . $importPath . '": circular reference detected.');
+                throw new FileImportException(
+                    \esc_html('Failed to import file "' . $importPath . '": circular reference detected.')
+                );
             }
 
             // grab referenced file & minify it (which may include importing
@@ -783,7 +785,9 @@ class CSS extends Minify
     {
         $result = preg_replace($pattern, $replacement, $subject);
         if ($result === null) {
-            throw PatternMatchException::fromLastError("Failed to replace with pattern '$pattern'");
+            throw PatternMatchException::fromLastError(
+                \esc_html("Failed to replace with pattern '$pattern'")
+            );
         }
 
         return $result;
diff --git a/src/Minify.php b/src/Minify.php
index ea3b295..cd3c5ed 100644
--- a/src/Minify.php
+++ b/src/Minify.php
@@ -132,7 +132,9 @@ abstract class Minify
 
             // check if we can read the file
             if (!$this->canImportFile($path)) {
-                throw new IOException('The file "' . $path . '" could not be opened for reading. Check if PHP has enough permissions.');
+                throw new IOException(
+                    \esc_html('The file "' . $path . '" could not be opened for reading. Check if PHP has enough permissions.')
+                );
             }
 
             $this->add($path);
@@ -241,6 +243,7 @@ abstract class Minify
 
         $this->writeToFile($handler, $content);
 
+        // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_fclose
         @fclose($handler);
     }
 
@@ -351,7 +354,7 @@ abstract class Minify
                 } else {
                     if ($matchResult === false) {
                         throw PatternMatchException::fromLastError(
-                            "Failed to match pattern '$pattern' at $processedOffset"
+                            \esc_html("Failed to match pattern '$pattern' at $processedOffset")
                         );
                     }
                     // if the pattern couldn't be matched, there's no point in
@@ -502,7 +505,8 @@ abstract class Minify
      */
     protected function canImportFile($path)
     {
-        $parsed = parse_url($path);
+        // phpcs:ignore WordPress.WP.AlternativeFunctions.parse_url_parse_url
+        $parsed = parse_url($path);
         if (
             // file is elsewhere
             isset($parsed['host'])
@@ -532,8 +535,11 @@ abstract class Minify
      */
     protected function openFileForWriting($path)
     {
+        // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_fopen
         if ($path === '' || ($handler = @fopen($path, 'w')) === false) {
-            throw new IOException('The file "' . $path . '" could not be opened for writing. Check if PHP has enough permissions.');
+            throw new IOException(
+                \esc_html('The file "' . $path . '" could not be opened for writing. Check if PHP has enough permissions.')
+            );
         }
 
         return $handler;
@@ -552,10 +558,13 @@ abstract class Minify
     {
         if (
             !is_resource($handler)
+            // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_fwrite
             || ($result = @fwrite($handler, $content)) === false
             || ($result < strlen($content))
         ) {
-            throw new IOException('The file "' . $path . '" could not be written to. Check your disk space and file permissions.');
+            throw new IOException(
+                \esc_html('The file "' . $path . '" could not be written to. Check your disk space and file permissions.')
+            );
         }
     }
 
