--- a/src/Wa72/Url/Url.php	2026-03-22 00:36:40.786920500 +0100
+++ b/src/Wa72/Url/Url.php	2026-03-22 00:36:40.796920500 +0100
@@ -31,7 +31,8 @@
         if ($this->is_protocol_relative()) {
             $url = 'http:' . $url;
         }
-        $urlo = parse_url($url);
+        // phpcs:ignore WordPress.WP.AlternativeFunctions.parse_url_parse_url
+        $urlo = parse_url($url);
         if (isset($urlo['scheme']) && !$this->is_protocol_relative()) {
             $this->scheme = strtolower($urlo['scheme']);
         }
